First version for the full Mender QA pipeline for GitLab

lluiscampos opened the pull request:
This commit introduces the first production version of the Mender QA
pipeline for GitLab CI.

Basically, what Jenkins runs through jenkins-yoctobuild-init-script.sh
has been ported to before_script on GitLab pipeline; while the clone of
all microservices that is specified for Jenkins in its configuration, is
done in the main script of GitLab pipeline.

Functionality wise, it reassembles Jenkins builds with the following
exceptions:
* Integration tests are run in parallel with acceptance tests
* There is no support for integration tests with precompiled client
* Upload of artifacts to S3 and Docker pushes (and therefore releases)
are not supported. This is a temporary workaround to not dirturbe the
development process.

The runs are slower than Jenkins due to missing optimizations:
* QA-79: KVM acceleration for qemu emulation
* QA-80: NFS mount for yocto builds

Implementaion detail. Three different Docker images are in use by the
differnt jobs:
* Init stage uses a simple Alpine Linux image
* Build and acceptance tests jobs use a Ubuntu based image, with Docker
as a service (DinD) to build the Docker images of the microservices. A
Linux user is created to run bitbake as non root.
* Integration tests jobs use a privilaged official DinD image to allow
composition of containers and their shared directories as the tests
require.

Changelog: Title

Signed-off-by: Lluis Campos <lluis.campos@northern.tech>

lluiscampos commented:
@mzedel @kacf Please have a close look to the Pipeline.

I am thinking to prepare a short meeting next Monday with the full team to go through it and resolve questions and concerns.

kacf commented on .gitlab-ci.yml:
```diff
@@ -1,37 +1,591 @@
+    - git init . && git remote add origin https://github.com/mendersoftware/poky || true
+    - git fetch && git checkout -f origin/${POKY_REV}
+    - echo POKY_REV_GIT_SHA=$(git rev-parse HEAD) > ${CI_PROJECT_DIR}/build_revisions.env
+    - git clone https://github.com/mendersoftware/meta-mender || true
```
Why all the `|| true`? Shouldn't it error if the clone fails?

lluiscampos replied:
The folders might already exist when the slaves are reused, so the clone will fail. If there is an error with the repo itself, we will still catch it in the next line, right?

Please review well also the next line, where I do the fetch pull/xxx/head into local branch pr. Will git call be robust when reusing slaves? I haven't verified it.

kacf replied:
Oh, I thought that since it's a container, it would always be fresh. But ok, I understand the approach then.

lluiscampos replied:
I thought so too... But experiments have showed otherwise.

What about this kind of checkouts: `(cd meta-mender && git fetch -u -f origin ${META_MENDER_REV}:pr && git checkout pr)`. Do they look good to you?

mzedel replied:
wouldn't a `rm -rf meta_mender && git clone -b ${META_MENDER_REV} --single-branch https://github.com/mendersoftware/meta-mender.git` also work? then we would reduce the 3 lines per project + speed up the cloning, by only focusing on the single-branch

- git clone https://github.com/mendersoftware/mender go/src/github.com/mendersoftware/mender || true
- (cd go/src/github.com/mendersoftware/mender && git fetch -u -f origin ${MENDER_REV}:pr && git checkout pr)

kacf replied:
Don't use single-branch. There is a lot of Git magic being used in various places that rely on a full repository. One example is the extremely long running Uboot compile test (~2h for a single test), which uses branches to find out whether the feature has been recently updated or not, and skipping based on this.

kacf commented on .gitlab-ci.yml:
```diff
@@ -1,37 +1,591 @@
+    - tar -czf /tmp/workspace.tar.gz .
+    - mv /tmp/workspace.tar.gz ${CI_PROJECT_DIR}/workspace.tar.gz
+  artifacts:
+    expire_in: 24h
```
Would it be good to have a bit more here? Unless it is extremely space consuming I would suggest maybe 2w or something. It's not uncommon for investigations to take a bit of time if you're interrupted by other things.

Same goes for other occurrences.

lluiscampos replied:
Yeah, I see your point.

It is 350MB and I am not sure how the artifact storage is billed from GitLab. @mzedel any input here?

mzedel replied:
If I understand the docs right, the limit per artifact and stage might be 2000mb and on Gitlab.com they can potentially live forever, while the default is 30days...
I don't mind increasing the expiry time until we are contacted about it, the only concern would be a "fair use" expiry date to be a good practice for an open source project.
Good point to support debugging with a bit more time and 2w should probably be good. If you agree, I'll adjust this across all my PRs.

lluiscampos replied:
Good, thanks for checking it. Let's make 2w the default for all our artifacts. I'll fix here

kacf commented on .gitlab-ci.yml:
```diff
@@ -1,37 +1,591 @@
+    - cp /sbin/debugfs /usr/bin/ || echo "debugfs not in /sbin/"
+    # sysstat monitoring suite
+    # collect cpu, load avg, memory and io usage every 2 secs forever
+    # use 'sadf' from sysstat to render the result file (~/sysstat.log) manually
```
`sadf`? Or `sar`?

lluiscampos replied:
TBH, I have no idea. That is copied as is in the `jenkins-yoctobuild-init-script.sh`

lluiscampos replied:
Should be `sar` :+1: I am fixing it in both palces

kacf commented on .gitlab-ci.yml:
```diff
@@ -1,37 +1,591 @@
+    - integration/extra/release_tool.py --set-version-of mender-conductor-enterprise --version pr
+    - integration/extra/release_tool.py --set-version-of tenantadm --version pr
+    - integration/extra/release_tool.py --set-version-of useradm --version pr
+    - integration/extra/release_tool.py --set-version-of mender-client-qemu --version pr
```
Should use a loop here, like:

```
for repo in `integration/extra/release_tool.py -l`; do
    integration/extra/release_tool.py --set-version-of $repo --version pr
done
```

This way we won't ever miss a version that we add later. And it will reveal if we have missed it anywhere else (there will be `pr` tag which can't be pulled).

lluiscampos replied:
Thanks! I'll rework

Side note: we could also have something like `--set-version-of -a --version pr`

lluiscampos replied:
@kacf Very good catch! I used `-l docker -a` options, so that we get the docker images and not the repos.

We were missing some of them (i.e. api-gateway, email-sender, mender-client-docker). And I think some of the failing integration tests are related to this.

lluiscampos commented:
> Really great work here @lluiscampos, nicely captured in Gitlab's CI format! I have a few comments, but it's just nitpicking.

Thanks :)

I am sure that once we start using it we will find improvements required. But is a good start, I think.

mzedel commented on .gitlab-ci.yml:
```diff
@@ -1,37 +1,592 @@
+      junit: results_qemux86_64_uefi_grub.xml
+  only:
+    variables:
+      # TODO: investigate why the expressions doesn't work in GitLab
```
apparently this was only recently merged: https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/27925 - should be part of the 12.0 or 12.1 release...

what might work (according to https://gitlab.com/gitlab-org/gitlab-ce/issues/27818#note_38463332): 
```
only:
  variables:
    - $RUN_INTEGRATION_TESTS == "true"
  variables:
    - $TEST_QEMUX86_64_UEFI_GRUB == "true"
```

lluiscampos replied:
Thanks! Added it and testing tonight...

lluiscampos replied:
It works :+1:

mzedel commented on .gitlab-ci.yml:
```diff
@@ -1,37 +1,592 @@
+  - docker:dind
+  tags:
+    - mender-qa-slave
+  before_script:
```
we should really put this into a base image as well as the https://github.com/mendersoftware/mender-test-containers - at least all the install only steps

lluiscampos replied:
Thanks. I'll create a JIRA task in the GitLab optimizations bucket. I have few of these that we can go through in the next QA meeting

lluiscampos commented:
IMO this is ready to go.

@kacf I think I have addressed all your comments. Have a second check and approve.

Some of the integration test failures are due to the TestDemoArtifact instability, but other failures have no known reason. But in any case they are not related to GitLab directly, and the sooner I get this into master the sooner I can collect stats for the builds.

lluiscampos commented:
Merging :)
